### PR TITLE
Issues/285 make RNA-seq functions reusable (resolves #287, resolves #283, resolves #BD2KGenomics/cgl-docker-lib#122, resolves #288, resolves #290, resolves #211)

### LINKED ISSUE #211: Review all asserts
An `assert` statement is used to catch bugs. We currently abuse them to validate inputs. Those abuses should be changed to raises of specific exceptions.

See comment by @fnothaft to the 1.0 PR.

https://github.com/BD2KGenomics/toil-scripts/pull/208#discussion_r57346105

### LINKED ISSUE #283: Remove bad RSEM post-processing from RNA-seq pipeline

### LINKED ISSUE #287: Add maxCores fix to RNA-seq / BWA pipelines
Currently Toil's `maxCores` argument doesn't really work since Docker tools are run with the cores value from `multiprocessing.cpu_count()`

### LINKED ISSUE #288: Move generic job functions for tools to common lib (RNA-seq)

### LINKED ISSUE #290: Improve readability of configs (exome/RNA)
Having comments on the same line as YAML inputs means once the user fills in options the formatting gets all yucko.

### PR DESCRIPTION
**Pull request by jvivian**
On June 2nd, I forgot about no periods in commits. I have undergone self-flagellation as an act of penitence for this indiscretion.

### CODE REVIEW

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.

**Comment by jvivian** [comment]
@hannes-ucsc Ready for review. I got an email last night from Brian and Josh about UCLA wanting to use my Docker container, for which this PR contains a fix that I need (#287).  I'll ask you at scrum today about maybe hotfixing the single commit into master so I can expedite getting this container to them if this PR isn't quickly reviewable.

File: src/toil_scripts/exome_variant_pipeline/README.md [hunk 1]
```diff
@@ -26,6 +26,15 @@ or `pip install --pre toil-scripts` for cutting edge development version.
 
 Type: `toil-exome` to get basic help menu and instructions
 
+Our software lead prefers installation of python projects into a virtualenv.
```
Comment (hannes-ucsc) [inline src/toil_scripts/exome_variant_pipeline/README.md]: Give me a break. This is not a mere preference of mine. It is a fact that not using virtualenvs increases the chance of versioning conflicts that can cause serious and hard to find bugs. I thought I had made that clear.

File: src/toil_scripts/rnaseq_cgl/README.md [hunk 2]
```diff
@@ -24,6 +24,15 @@ or `pip install --pre toil-scripts` for cutting edge development version.
 
 Type: `toil-rnaseq` to get basic help menu and instructions
 
+Our software lead prefers installation of python projects into a virtualenv.
```
Comment (hannes-ucsc) [inline src/toil_scripts/rnaseq_cgl/README.md]: Same here.

File: src/toil_scripts/exome_variant_pipeline/exome_variant_pipeline.py [hunk 3]
```diff
@@ -607,25 +613,39 @@ def generate_config():
...
-    run-muse: true          # Optional: If true, will run MuSe to do mutation calls
-    preprocessing: true     # Optional: If true, will perform indel realignment and base quality score recalibration\n
-    output-dir:             # Optional: Provide a full path to where results will appear\n
-    s3-dir:                 # Optional: Provide an s3 path (s3://bucket/dir) where results will appear\n
-    ssec:                   # Optional: Provide a full path to a 32-byte key used for SSE-C Encryption in Amazon\n
-    gtkey:                  # Optional: Provide a full path to a CGHub Key used to access GNOS hosted data\n
-    ci-test:                # Optional: If true, uses resource requirements appropriate for continuous integration\n
+    # Required: URL to reference genome
+    # Example: s3://cgl-pipeline-inputs/variant_hg19/hg19.fa
+    reference: \n
```
Comment (hannes-ucsc) [inline src/toil_scripts/exome_variant_pipeline/exome_variant_pipeline.py]: What would be the impact of 

```
# Required: URL to reference genome
reference: s3://cgl-pipeline-inputs/variant_hg19/hg19.fa
```

Comment (hannes-ucsc) [inline src/toil_scripts/exome_variant_pipeline/exome_variant_pipeline.py]: And I would get rid of the `\n` to generate empty lines in the output without having them in the code. Either put empty lines in the code or remove them altogether.

**Comment by hannes-ucsc** [comment]
I did some spot-checking but in all honesty, I don't have the time to do a in-depth review.

I have to admit that I am slightly offended by the virtualenv comment.

File: src/toil_scripts/exome_variant_pipeline/README.md (hunk 1 above)
Comment (jvivian) [inline src/toil_scripts/exome_variant_pipeline/README.md]: This was in *no way* meant to be a dig, merely an explanation of why I'm asking the user to install it into a virtualenv, something a user may never have seen as part of pip installing a package.  I'm not disagreeing with your point whatsoever. I'll go ahead and substitute it with your more substantiative explanation.

File: src/toil_scripts/exome_variant_pipeline/exome_variant_pipeline.py (hunk 3 above)
Comment (jvivian) [inline src/toil_scripts/exome_variant_pipeline/exome_variant_pipeline.py]: I like your first suggestion — streamlines running these for our production team. 

I like the aesthetic of the spacing between lines, so I'll probably add it directly into the code.

**Comment by jvivian** [comment]
Jenkins, test this please.

**Comment by jvivian** [comment]
Jenkins, test this please.

**Comment by jvivian** [comment]
Keep getting:
```
Submitting dist/toil-scripts-2.0a3.dev231.tar.gz to https://pypi.python.org/pypi
Upload failed (500): Internal Server Error
error: Upload failed (500): Internal Server Error
```

Jenkins, test this please.

**Comment by jvivian** [comment]
Jenkins, test this please.

**Comment by jvivian** [comment]
Finally, thank you Jenkins.

@hannes-ucsc — Ready for review.  I apologize for the size of this PR and will do my best to keep PRs single-issue as best as I can in the future.

This PR boils down to:  

- Restructuring tool job functions for RNA-seq to **toil_scripts/tools/**
- MaxCores fix to propagate maxCores to tools
- RSEM fix to remove bad output files

Everything else is just README, docstring, or aesthetic code cleanup.

**Comment by hannes-ucsc** [comment]
Let's try and keep the commit titles 72 characters or shorter.

**Pull request merged**